File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1290.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

function addLast(head: ListNode, val: number): void {
    let currNode: ListNode | null = head;
    while (currNode.next != null) {
        currNode = currNode.next;
    }
    currNode.next = new ListNode(val);
}

function printList(head: ListNode): void {
    let currNode: ListNode | null = head;
    let result: string = "";
    while (currNode != null) {
        result += `${currNode.val} --> `;
        currNode = currNode.next;
    }
    result += "Null";
    console.log(result);
}

function getDecimalValue(head: ListNode | null): number {
    let res: number = 0;
    while (head != null) {
        if (head.val === 0) {
            res = res << 1;
        } else {
            res = (res << 1) + 1;
        }
        head = head.next;
    }
    return res;
};

let head: ListNode = new ListNode(1);
addLast(head, 0);
addLast(head, 1);
printList(head);
console.log(`Decimal value: ${getDecimalValue(head)}`);

export { };

File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1290.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class Leetcode1290 {
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        addLast(head, 0);
        addLast(head, 1);
        printList(head);
        System.out.println("The Decimal value: " + getDecimalValue(head));
    }

    private static void addLast(ListNode head, int val) {
        ListNode currNode = head;
        while (currNode.next != null) {
            currNode = currNode.next;
        }
        currNode.next = new ListNode(val);
    }

    private static void printList(ListNode head) {
        ListNode currNode = head;
        while (currNode != null) {
            System.out.print(currNode.val + " --> ");
            currNode = currNode.next;
        }
        System.out.println("NULL");
    }

    private static int getDecimalValue(ListNode head) {
        int num = 0;
        while (head != null) {
            if (head.val == 0) {
                num = num << 1;
            } else {
                num = (num << 1) + 1;
            }
            head = head.next;
        }
        return num;
    }
}